pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol

```diff
@@ -0,0 +1,285 @@
+     * @dev If an upper or lower bound value is zero, it means there is no circuit breaker in that direction for the
+     * given token.
+     * @param circuitBreakerState - The bytes32 state of the token of interest.
+     */
+    function getCircuitBreakerFields(bytes32 circuitBreakerState) internal pure returns (CircuitBreakerParams memory) {
+        return
+            CircuitBreakerParams({
+                referenceBptPrice: circuitBreakerState.decodeUint(
+                    _REFERENCE_BPT_PRICE_OFFSET,
+                    _REFERENCE_BPT_PRICE_WIDTH
+                ),
+                referenceWeightComplement: circuitBreakerState
+                    .decodeUint(_REFERENCE_WEIGHT_COMPLEMENT_OFFSET, _REFERENCE_WEIGHT_COMPLEMENT_WIDTH)
+                    .decompress(_REFERENCE_WEIGHT_COMPLEMENT_WIDTH, _MAX_BOUND_PERCENTAGE),
+                lowerBoundPercentage: circuitBreakerState
+                    .decodeUint(_LOWER_BOUND_PCT_OFFSET, _BOUND_PERCENTAGE_WIDTH)
+                    .decompress(_BOUND_PERCENTAGE_WIDTH, _MAX_BOUND_PERCENTAGE),
+                upperBoundPercentage: circuitBreakerState
+                    .decodeUint(_UPPER_BOUND_PCT_OFFSET, _BOUND_PERCENTAGE_WIDTH)
+                    .decompress(_BOUND_PERCENTAGE_WIDTH, _MAX_BOUND_PERCENTAGE)
+            });
+    }
+
+    /**
+     * @notice Returns the dynamic lower and upper BPT price bounds for a given token, at the current weight.
+     * @dev The current BPT price of the token can be directly compared to these values, to determine whether
+     * the circuit breaker has tripped. If a bound is 0, it means there is no circuit breaker in that direction
+     * for this token: there might be a lower bound, but no upper bound. If the current BPT price is less than
+     * the lower bound, or greater than the non-zero upper bound, the transaction should revert.
+     *
+     * These BPT price bounds are dynamically calculated using the conversion ratios. In general:
+     * lower/upper BPT price bound = referenceBptPrice * "conversion ratio". The conversion ratio is given as
+     * (boundary percentage)**(currentWeightComplement).
+     *
+     * If the value of the weight complement has not changed, we can use the reference conversion ratios stored
+     * when the breaker was set. Otherwise, we need to calculate them.
+     *
+     * As described in the general comments above, the weight complement calculation attempts to isolate changes
+     * in the balance due to arbitrageurs responding to external prices, from internal price changes caused by an
+     * ongoing weight update, or changes to the pool composition. There is a non-linear relationship between "spot"
+     * price changes and BPT price changes. This calculation transforms one into the other.
+     *
+     * @param circuitBreakerState - The bytes32 state of the token of interest.
+     * @param currentWeightComplement - The complement of this token's weight: (_denormWeightSum - token weight)
```

Comment:
Same comment re. what the complement is. I don't see how it could be sum - weight, as that is an arbitrary value.